src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs

```diff
@@ -0,0 +1,162 @@
+// <copyright file="RequestTelemetryStateTracker.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.ServiceModel.Channels;
+using System.Threading;
+
+namespace OpenTelemetry.Instrumentation.Wcf.Implementation;
+
+internal class RequestTelemetryStateTracker : IDisposable
+{
+    private readonly OrderedDictionary outstandingRequestStates = new OrderedDictionary();
+    private readonly object sync = new object();
+    private Timer timer;
+
+    private long configuredTimeoutInMs;
+
+    public RequestTelemetryStateTracker(TimeSpan timeout)
+    {
+        this.configuredTimeoutInMs = (long)timeout.TotalMilliseconds;
+    }
+
+    public event EventHandler<RequestTelemetryState> TelemetryStateTimedOut;
+
+    public void PushTelemetryState(Message request, RequestTelemetryState telemetryState)
+    {
+        var messageId = request?.Headers.MessageId?.ToString();
+        if (messageId != null)
+        {
+            var expiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + this.configuredTimeoutInMs;
+            var entry = new Entry(telemetryState, expiresAt);
+            lock (this.sync)
+            {
+                this.outstandingRequestStates.Add(messageId, entry);
```

Comment:
There is an obscure feature where you can cast an channel to an `IContextChannel` and set the `OperationTimeout` property to a different value, subsequent calls will use that timeout. You can't presume later requests will have a later deadline. You need to keep track of the current timers deadline and compare the new deadline to see if the timer needs to fire sooner.

Reply:
Ahhh shucks.  Y'know, I actually initially implemented it with the dynamic timeout functionality - rather than a timer/dict per channel I had a single global one for all channels (where each channel could have different timeouts configured)....but I couldn't find a good data structure to use.  Do you know of an available data structure which allows you to hash on 1 key (the message id) but sort by another (the deadline)?  In my initial implementation I ended up kinda implementing my own, and it came out a little gross so I figured a dict per-channel ordered by insertion order was cleaner....but if it needs to handle dynamic timeouts within a single channel then I guess that's not an option.